packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,190 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { ExtractionClient, ExtractorState, MappingsClient, REPORTING_BASE_PATH, ReportsClient } from "@itwin/insights-client";
+import type { ReportMapping } from "@itwin/insights-client";
+import { generateUrl, handleError } from "./utils";
+import type { CreateTypeFromInterface } from "./utils";
+import type {
+  GetSingleIModelParams,
+  IModelsClientOptions,
+} from "@itwin/imodels-client-management";
+import { Constants, IModelsClient } from "@itwin/imodels-client-management";
+import { AccessTokenAdapter } from "@itwin/imodels-access-frontend";
+import type { ReportsApiConfig } from "../context/ReportsApiConfigContext";
+import { ExtractionStates } from "./ExtractionStatus";
+
+export type ReportMappingType = CreateTypeFromInterface<ReportMapping>;
+
+export type ReportMappingAndMapping = ReportMappingType & {
+  mappingName: string;
+  mappingDescription: string;
+  iModelName: string;
+};
+
+export default class BulkExtractor {
+  private _reportRunIds = new Map<string, string[]>();
+  private _apiConfig: ReportsApiConfig;
+
+  constructor(apiConfig: ReportsApiConfig) {
+    this._apiConfig = apiConfig;
+  }
+
+  public async getStates(reportIds: string[]): Promise<Map<string, ExtractionStates>> {
+    const extractionClientApi = new ExtractionClient(
+      generateUrl(REPORTING_BASE_PATH, this._apiConfig.baseUrl)
+    );
+    const accessToken = await this._apiConfig.getAccessToken();
+
+    const stateByReportId = new Map<string, ExtractionStates>();
+    const stateByRunId = new Map<string, ExtractorState>();
+
+    for (const reportId of reportIds) {
+      const runs = this._reportRunIds.get(reportId);
+      if (!runs) {
+        stateByReportId.set(reportId, ExtractionStates.None);
+        continue;
+      }
+
+      const states: ExtractorState[] = [];
+      for (const runId of runs) {
+        const state = stateByRunId.get(runId);
+        if (state) {
+          states.push(state);
+        } else {
+          const runState = await this.getSingleState(runId, extractionClientApi, accessToken);
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = this.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  public clearJobs() {
+    this._reportRunIds = new Map<string, string[]>();
+  }
+
+  private getFinalState(states: ExtractorState[]) {
+    const responseState = states.includes(ExtractorState.Failed) ? ExtractionStates.Failed :
```

Comment:
There is no benefit of having `responseState` variable.

Replace it with `if`s that return immediately if the condition is true.

If no conditions are met - return `Failed`.

Reply:
`responseState` variable has been removed. State is returned immediately